### PR TITLE
feat: add `unlock_account` function to `ProviderAPI`

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did
This PR adds a feature that allows working with "impersonated accounts", which is meant to be sublclassed by providers that work in forked environments that support account impersonation (the ability to submit transactions for accounts you do not have the private key for)

fixes: #488

### How I did it
- [x] Add `unlock_account` method (which raises `NotImplementedError` by default) to `ProviderAPI`
- [x] Add support to `AccountManager.__getitem__` for loading impersonated accounts
- [x] Modify `test_accounts` fixture to allow loading impersonated accounts (currently just a `List`)
- [x] Support skipping signatures if on a test provider that supports it
- [x] Update docs to describe this feature

### How to verify it

Use this PR: https://github.com/ApeWorX/ape-hardhat/pull/37

Load a console in a forked mode environment, and use `impersonated_account = accounts["<some address or ENS>"]`

Make a transaction using the impersonated account
Make a normal transaction (to regression test).

### Checklist

- [x] All changes are completed
- [ ] New test cases have been added
- [x] Documentation has been updated

### CODE REVIEW

**Comment by fubuloubu** [comment]
> Support skipping signatures if on a test provider that supports it

@unparalleled-js may need help with this part, but the rest works!

```py
$ ape console --network fantom:opera-fork
INFO: Starting 'Hardhat node' process.
WARNING: Connecting Geth plugin to non-Geth network 'go-opera'.
WARNING: Upstream network has mismatching genesis block. This could be an issue with hardhat.

In [1]: rando = accounts["0x0024aA72eb2cB8110A137881E5398B8445674741"]

In [2]: rando.balance
Out[2]: 9839395855413250000
```

File: src/ape/api/providers.py [hunk 1]
```diff
@@ -574,6 +574,18 @@ def mine(self, num_blocks: int = 1):
...
+        """
+        Ask the provider to allow an address to submit transactions without validating
+        signatures. This feature is intended to be subclassed by a
+        :class:`~ape.api.providers.TestProviderAPI` so that during a fork-mode test,
+        a transaction can be submitted by an arbitrary account or contract without a private key.
+
+        Raises:
+            NotImplementedError: When this provider does not support unlocking an account.
+        """
+        raise NotImplementedError("`unlock_account` is not implemented by this provider")
```
Comment (unparalleled-js) [inline src/ape/api/providers.py:587]: You can use the `@raises_not_implemented` decorator

File: src/ape/managers/accounts.py [hunk 2]
```diff
@@ -1,12 +1,81 @@
...
+    @__getitem__.register
+    def __getitem_int(self, account_id: int):
+        for idx, account in enumerate(self.accounts):
+            if account_id == idx:
+                return account
+
+        raise IndexError(f"No account at index '{account_id}'.")
+
+    @__getitem__.register
+    def __getitem_str(self, account_str: str):
```
Comment (unparalleled-js) [inline src/ape/managers/accounts.py:58]: It could be on another PR, but we could support aliases here so you can load our own accounts in a test environment and automatically have funding.

**Review by unparalleled-js — COMMENTED** [review]
Nice!

We can also throw a sentence in the `project` user-guide (section Testing -> Fixtures -> Accounts) about how if you are using a mainnet-fork provider that supports impersonation, you can load accounts by address that way.

File: src/ape/api/providers.py [hunk 3]
```diff
@@ -574,6 +574,18 @@ def mine(self, num_blocks: int = 1):
             NotImplementedError: Unless overridden.
         """
 
+    def unlock_account(self, address: AddressType) -> bool:
```
[the remark this answers is not in the document]
Comment (unparalleled-js) [reply src/ape/api/providers.py:577]: Agh, I just realized adding the `abstractmethod` to test account api would be a breaking change... 
We can punt on that and try to get rid of `TestAccountAPI`

File: src/ape/managers/accounts.py [hunk 4]
```diff
@@ -1,12 +1,81 @@
...
 
+class ImpersonatedAccount(AccountAPI):
+    raw_address: AddressType
+
+    @property
+    def address(self) -> AddressType:
+        return self.raw_address
+
+    def sign_message(self, msg: SignableMessage) -> Optional[MessageSignature]:
+        raise NotImplementedError("This account cannot sign messages")
```
[the remark this answers is not in the document]
Comment (fubuloubu) [reply src/ape/managers/accounts.py:19]: Yes, you're totally right, I have no idea why I did this lol

[the remark this answers is not in the document]
Comment (fubuloubu) [reply src/ape/managers/accounts.py:19]:
```suggestion
        return None
```

**Comment by unparalleled-js** [comment]
This will help with https://github.com/ApeWorX/ape/issues/232 too

File: src/ape/api/accounts.py [hunk 5]
```diff
@@ -384,3 +384,25 @@ class TestAccountAPI(AccountAPI):
...
+    def address(self) -> AddressType:
+        return self.raw_address
+
+    def sign_message(self, msg: SignableMessage) -> Optional[MessageSignature]:
+        raise NotImplementedError("This account cannot sign messages")
+
+    def sign_transaction(self, txn: TransactionAPI) -> Optional[TransactionSignature]:
+        return None
+
+    def call(self, txn: TransactionAPI, send_everything: bool = False) -> ReceiptAPI:
```
Comment (fubuloubu) [inline src/ape/api/accounts.py:406]: Oh nice, well this definitely works!

**Comment by fubuloubu** [comment]
Is there a way to test this feature?

File: src/ape/contracts/base.py [hunk 6]
```diff
@@ -56,9 +51,10 @@ def serialize_transaction(self, *args, **kwargs) -> TransactionAPI:
...
     def __call__(self, *args, **kwargs) -> ReceiptAPI:
-        if "sender" in kwargs:
+        txn = self.serialize_transaction(*args, **kwargs)
+
+        if "sender" in kwargs and isinstance(kwargs["sender"], AccountAPI):
             sender = kwargs["sender"]
-            txn = self.serialize_transaction(*args, **kwargs)
             return sender.call(txn)
 
         txn = self.serialize_transaction(*args, **kwargs)
```
Comment (unparalleled-js) [inline src/ape/contracts/base.py:60]: - [ ] don't need to call this again here, use the one hoisted up

File: src/ape/managers/accounts.py [hunk 7]
```diff
@@ -1,12 +1,66 @@
...
-from ape.api.accounts import AccountAPI, AccountContainerAPI, TestAccountAPI
+from ape.api.accounts import AccountAPI, AccountContainerAPI, ImpersonatedAccount, TestAccountAPI
 from ape.types import AddressType
-from ape.utils import cached_property, singledispatchmethod
+from ape.utils import ManagerAccessMixin, cached_property, singledispatchmethod
 
 from .base import BaseManager
 
 
+class TestAccountContainer(list, ManagerAccessMixin):
```
Comment (unparalleled-js) [inline src/ape/managers/accounts.py:10]: Look into this being the actual `TestAccountContainer` registered OR possibly rename

**Review by unparalleled-js — COMMENTED** [review]
Address feedback from @NotPeopling2day

File: src/ape/managers/accounts.py (hunk 7 above)
Comment (unparalleled-js) [reply src/ape/managers/accounts.py:10]: This is actually like a container of containers...

The thing is... What if different ecosystems register their own test accounts (meaning they have their own account containers that yield them)? You would get all those accounts here, so 0-19 are my ethereum test accounts and 20-29 are my starknet ones, etc.

So I am leaning on **not** moving this to `ape-test.TestAccountContainer`

Comment (unparalleled-js) [reply src/ape/managers/accounts.py]: Renamed to `TestAccountManager`

**Comment by unparalleled-js** [comment]
- [ ] Handle being able to unlock your keyfile accounts

**Review by NotPeopling2day — APPROVED** [review]
